### ISSUE TITLE
Anti-redirect scriptlet for vk.com

### ISSUE THREAD

**Issue by dimisa-RUAdList** [uBlockOrigin/uBlock-issues#2531]
### Prerequisites

- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uBlock-issues/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] The issue is not present after disabling uBO in the browser.
- [X] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I am reporting is not normal behavior.

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] using a new, unmodified browser profile.

### Description

The site vk.com is one of the most visited Russian sites. All external links on this site are done through a redirect.

There is an idea to create a scriptlet with a name like "anti-redirect" based on this script: https://greasyfork.org/en/scripts/395977-vk-com-del-redir/code

### A specific URL where the issue occurs.

https://vk.com/amedia.online

### Steps to Reproduce

undefined

### Expected behavior

undefined

### Actual behavior

undefined

### Configuration

uBlock Origin development build 1.47.5b2

**Comment by gorhill** [comment]
Your request is the equivalent of asking uBO to deal with redirect link, something which I have declined in the past, <https://github.com/uBlockOrigin/uBlock-issues/issues/1784>. [Clear URLs](https://addons.mozilla.org/en-US/firefox/addon/clearurls/) is the right tool for this.

**Comment by dimisa-RUAdList** [comment]
The thing is, ClearURLs doesn't solve the problem on this site. The transition is still done via `vk.com/away.php`.

**Comment by gorhill** [comment]
If it doesn't work with ClearURLs then maybe it's best to report the issue on [their repo](https://github.com/ClearURLs/Addon/issues)?

**Comment by dimisa-RUAdList** [comment]
Thanks, but I thought it might be possible to solve this problem with uBO. At the moment, almost all such tasks on Russian sites are solved by the Counters filter. However, the current features of uBO do not allow me to do the same for `vk.com`. Maybe you still look at the [script code](https://greasyfork.org/en/scripts/395977-vk-com-del-redir/code)?

**Comment by gorhill** [comment]
I prefer to decline, as explained in these similar requests:

- https://github.com/uBlockOrigin/uBlock-issues/issues/1784
- https://github.com/gorhill/uBlock/issues/3276
- https://github.com/uBlock-LLC/uBlock/issues/1009

Adding a scriptlet for this one particular case, beside that `vk.com` could change at any time (further adding burden here), it opens the door for all such requests in the future. This is best left to a dedicated extension, or even more simply, just a matter of using a user script extension.

**Comment by dimisa-RUAdList** [comment]
@MasterKia 
Thank you, I have not yet given up hope of creating a solution using the default tools of uBO.

@gorhill 
I carefully reviewed all the rejected cases, including my own.

I think it is possible to create a solution that will be adequate and useful for many cases, although not for all.

I propose to create a scriptlet to modify the value of the href attribute. Its only function will be to replace the original href value with its text value.

Universal. No complicated and risky redirects, no additions, absolutely transparent, using only what already exists in the original layout.

Example: `https://vk.com/amedia.online`
<details>
<summary>Screenshot(s)</summary>

![vkhref](https://user-images.githubusercontent.com/20126984/224003869-d1d58d3c-42e0-4de8-b3d3-1074b7d56186.jpg)

</details>

Target: `href="/away.php?to=https%3A%2F%2Famedia.online%2F&cc_key="` ~> `https://amedia.online/`

Solution: `vk.com##+js(href-to-text, [href^="/away.php?to="])`

**Comment by dimisa-RUAdList** [comment]
@uBlock-user 
Thank you, but no. If I add such a rule to the filters, it will only be a terrible irritant. This is not a solution. This is some kind of sophisticated way to motivate users to remove uBO.

For now, I'd rather wait for a response from Gorhill.

**Comment by dimisa-RUAdList** [comment]
@uBlock-user 
Thanks for the clarification, but I'm aware of how Strict Blocking works.

However, its infrequent triggers are not at all the same as a constant trigger on every external transition on one of the most visited Russian sites.

Do not be offended, but I have no purpose to continue the discussion of the solution you proposed. I'm just waiting for a response from Gorhill.

**Comment by gorhill** [comment]
I added a `href-from-text` scriptlet as suggested. I did not need much convincing for this idea because I went through the same idea earlier this week with the `t.co` debacle -- where I envisioned that it would have been nice if uBO replaced the `t.co` link with the text shown to user which is the actual URL -- in which case users of uBO would have been unaffected by the debacle.

Having more than one use case for such scriptlet, especially both from high traffic sites, is enough from me to want to try this -- let's just consider it experimental at this point in case there are serious issues with using such scriptlets identified later.

So for your use case, this filter works:

    vk.com##+js(href-from-text, a[href^="/away.php?to="])

For Twitter, this works:

    twitter.com##+js(href-from-text, a[href^="https://t.co/"])

**Comment by gorhill** [comment]
That case is Youtube warning that you are leaving the site, so it's not done silently, and because of this I would rather that we do not bypass these warnings.

**Comment by gorhill** [comment]
Oh ok, I didn't realize this. Not sure if the issue with the event handler can be fixed, this will need some research.

**Comment by dimisa-RUAdList** [comment]
@gorhill 
I have found that in some cases, the link given in the text is shortened, and then the rule creates an invalid link.

The full link is contained in the `title` attribute.

Example: `https://vk.com/amedia.online`
<details>
<summary>Screenshot(s)</summary>

![hreffromtext](https://user-images.githubusercontent.com/20126984/224092205-381c014c-3ebd-4fd7-973b-83e1a5dca48f.jpg)

</details>

It looks like the scriptlet should be more flexible to be able to specify where exactly the link should come from.

Then the rules will be something like this:
`vk.com##+js(href-from-text, a[href^="/away.php?to="]:not([title]), text)`
`vk.com##+js(href-from-text, a[href^="/away.php?to="][title], [title])`

**Comment by gorhill** [comment]
Change in next dev build, however I renamed `href-from-text` to `href-sanitizer`. You can leave the second argument out to extract from text content (default).

**Comment by dimisa-RUAdList** [comment]
@gorhill 
Yes, everything is great now! The vast majority of links have become direct.

**Comment by dimisa-RUAdList** [comment]
@gorhill 
Sometimes a direct link is contained in an attribute on a child element.

Example: `https://vk.com/vokrugsveta_vk`
<details>
<summary>Screenshot(s)</summary>

![hreffromtext_child](https://user-images.githubusercontent.com/20126984/224164869-b84376d0-b897-400f-ab9a-0f9669530e9c.jpg)

</details>

Requires the ability to point to it: `vk.com##+js(href-sanitizer, a[class="media_link__media"][href^="/away.php?to="], a[class="media_link__media"] > [data-link-url])`

**Comment by dimisa-RUAdList** [comment]
@gorhill 
Can we expect the scriptlet to improve in terms of pointing to links contained in child elements?

**Comment by krystian3w** [comment]
Maybe failed on:

`https://twitter.com/Post_Courier/status/1121677734940332032` but I can not improve like on VK.

**Comment by gorhill** [comment]
Rught, it's not doing well on Twitter, maybe something changed because when I looked at it during the debacle, I am pretty sure all links where shown, now they are no longer shown, there is just a "card" with no information whatsoever about the actual link.

**Comment by Emi-TheDhamphirInLoveUnderTheFrozenStar** [comment]
@gorhill look at the link in the devtools, it is an `http`, so it is working as intended by you, only sanitizing https links, because for some reason many links in Twitter are http, so for example it will work great here: `https://twitter.com/lopatonok/status/1634326102947815424` because it is https.

Quickly allowing it to sanitize http links with `if ( /^https?:\/\/./.test(text) === false ) { return ''; }` will work.

![image](https://user-images.githubusercontent.com/122518587/224888287-2870fd73-31c5-47f2-923a-e0fc79b66e2a.png)

**Comment by stephenhawk8054** [comment]
Is it because the link in the above twitter link is `http:` instead of `https:`? When I change the matching line

```js
if ( /^https:\/\/./.test(text) === false ) { return ''; }
```

to

```js
if ( /^https?:\/\/./.test(text) === false ) { return ''; }
```

it works for me.

**Comment by v1gnesh** [comment]
Firstly, thank you for the tireless work from everyone involved in making the internet habitable.

I see that `t.co` thing doesn't work when the links show up in a tweet with the preview card.
Example: https://twitter.com/VoltronData/status/1641145746639421442

The actual URL isn't in the text of the tweet, because the link preview card takes its place.
Can this be made to work, by any chance?

**Comment by v1gnesh** [comment]
Ok :(, thanks for responding!

**Comment by gorhill** [comment]
That's just a domain name, that's not the actual destination URL.

**Comment by Yuki2718** [comment]
@gorhill Will it be possible to replace with the part of href attribute? For example, Android version of `https://www.mozilla.org/ja/firefox/browsers/mobile/android/` on my locale has this href: `https://app.adjust.com/2uo1qc?redirect=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dorg.mozilla.firefox&campaign=www.mozilla.org&adgroup=mobile-android-page` so the destination is included as the redirect parameter. I tried `mozilla.org##+js(href-sanitizer, #playStoreLink-primary, [data-mozillaonline-link])` but somehow `[data-mozillaonline-link]` is a wrong url and thus doesn't work. Maybe something like `mozilla.org##+js(href-sanitizer, #playStoreLink-primary, [href], ?redirect=)` to specify the starging point of the desired text, or regex to more exactly specify part of the text? I see this is a differnt form of [$queryjump](https://github.com/uBlockOrigin/uBlock-issues/issues/760#issuecomment-720080024).

**Comment by stephenhawk8054** [comment]
I think integrating it to some future `$queryjump` is better? To reduce the complexity of the scriptlet.

**Comment by Yuki2718** [comment]
`adjust.com` strict-blocking is a major annoyance at least in Japan, I received complaints from user and I myself occasionally come across this. Couldn't be happier if can be fixed earlier.

**Comment by stephenhawk8054** [comment]
Yeah, I also thought a [redirection like `$queryjump` idea that's compatible to strict-blocking would help reduce complaints](https://github.com/uBlockOrigin/uAssets/commit/2cb318118818088e08510d64840b8cdd0dc77003#commitcomment-109325362). Or it would help many cases of [email tracking like `awstrack.com`](https://github.com/uBlockOrigin/uAssets/issues/10585).

Or we can extend to some other strict-blocking filter without [worrying about troubles in large sites](https://github.com/uBlockOrigin/uAssets/commit/2cb318118818088e08510d64840b8cdd0dc77003#commitcomment-109339385).

But I think this needs another dedicated thread like `removeparam`.

**Comment by krystian3w** [comment]
https://github.com/uBlockOrigin/uBlock-issues/issues/2531#issuecomment-1512373333 - this can be added for users with European/American system/browser configurations:

https://user-images.githubusercontent.com/35370833/232686106-db8964bd-c6a6-4ad4-b514-769675cec983.png (works too on PC Firefox)

There will be a minimal reduction in reports, unless you need it in the Japanese AdGuard list instead of in a supplement https://github.com/Yuki2718/adblock2/blob/main/japanese/jpfp-ub.txt.

**Comment by Yuki2718** [comment]
> There will be a minimal reduction in reports, unless you need it in the Japanese AdGuard list instead of in a supplement

IDK what you mean.

**Comment by krystian3w** [comment]
What I'm trying to say is that you can already add this first filter (`mozilla.org##+js(href-sanitizer, #playStoreLink-primary, [data-mozillaonline-link])`) to the AdGuard JP "database" or your private project - in order to reduce complaints on e.g. Twitter (with non-normal configured PC in Japan - e.g. user froced Polish in settings due japanese is his second language e.g. Emil/Emma from YouTube), about not being able to download Firefox because Mozilla used ugly tracking links.

**Comment by Yuki2718** [comment]
>  I tried `mozilla.org##+js(href-sanitizer, #playStoreLink-primary, [data-mozillaonline-link])` but somehow `[data-mozillaonline-link]` is a wrong url and thus doesn't work.

**Comment by mapx-** [comment]
@gorhill 

html_sanitizer cannot be used on links like

`https://www.jpvhub.com/download?link=vidoza.net/lfz41z8271nw.html`

just because is missing `https://` ?

test page:
`https://www.jpvhub.com/jp/9edgp69d8k/jav/%E6%9C%89%E4%BF%AE%E6%AD%A3/first-impression-azumi-ipz-094`

    jpvhub.com##+js(href-sanitizer, a[href^="/download?link="])

**Comment by krystian3w** [comment]
Test latest idea instead skip parameter:

```adb
jpvhub.com##+js(href-sanitizer, a[href^="/download?link="], ?link)

```

**Comment by gorhill** [comment]
> Just because is missing `https://` ?

Yes that's the issue (if you use `?link` extra parameter). There is not way for the scriptlet to figure out whether what follows `?link=` is a relative or absolute URL. To handle such case we would need a trusted scriptlet with more power.

**Comment by dimisa-RUAdList** [comment]
How to remove such a redirect? ~> `https://yap.ru/forum2/topic2625123.html`
<details>
<summary>Screenshot(s)</summary>

![Yap](https://github.com/uBlockOrigin/uBlock-issues/assets/20126984/1ab51257-6e49-4dee-86f2-c7b1e66e8b82)

</details>

**Comment by krystian3w** [comment]
Do not like peppers?

`https://www.pepper.pl/promocje/zewnetrzny-dysk-ssd-crucial-x9-pro-1tb-oraz-inne-produkty-tej-marki-750379`

fails:

```adb
! chollometro.com,dealabs.com,hotukdeals.com,mydealz.de,pepper.com,pepper.pl,pepper.it,pepperdeals.se,preisjaeger.at,promodescuentos.com
chollometro.com,dealabs.com,hotukdeals.com,mydealz.de,pepper.com,pepper.pl,pepper.it,pepperdeals.se,preisjaeger.at,promodescuentos.com##+js(href-sanitizer, a[href*="https://www.pepper.pl/visit/"][title^="http"]. [title])
```

**Comment by Yuki2718** [comment]
Is it possible to add an argument to reverse the current behavior for `?` and also support `#` i.e. instead of using text after `?`, cut down text after specific symbol and use the rest href? This will solve https://github.com/DandelionSprout/adfilt/discussions/163#discussioncomment-8248048 where links in "More from WIRED" are like `https://www.wired.com/story/tech-layoffs-2024-amazon-google-discord-twitch/#intcid=_wired-bottom-recirc-version3_dd2b801a-5abd-47bf-b8d8-f12a1b71d254_roberta-similarity1`. Or it requires totally differnt solution other than `uritransfer`? I think `uritransfer` is overkill for this.

**Comment by stephenhawk8054** [comment]
About twitter `t.co` redirect, can anyone test if this filter works?

```adb
twitter.com,~platform.twitter.com##+js(trusted-replace-xhr-response, '/,"expanded_url":"([^"]+)","url":"[^"]+"/g', ',"expanded_url":"$1","url":"$1"', /api/graphql)
```

**Comment by Yuki2718** [comment]
It's working.

**Comment by D4niloMR** [comment]
On my end the matching url is `https://api.twitter.com/graphql/7xflPyRiUxGVbJd4uWmbfg/TweetResultByRestId?variables=[...]`

<details><summary>Screenshot</summary>

![image](https://github.com/uBlockOrigin/uBlock-issues/assets/70459964/09b1f43a-80cf-4ade-9738-a428bf989f9a)


</details>

**Comment by Yuki2718** [comment]
The rule stopped working.

**Comment by stephenhawk8054** [comment]
@Yuki2718 Currently I'm testing on just Firefox. Can you check again?

**Comment by Yuki2718** [comment]
I was wrong, the rule is still working for usual links. It doesn't work only for card links with images like one in `https://twitter.com/imabarizine/status/1784740979209355751`.

**Comment by stephenhawk8054** [comment]
Yeah, the card is quite hard, I don't know how to address it

**Comment by stephenhawk8054** [comment]
Hmm... From this link: https://help.twitter.com/en/using-x/url-shortener , looks like `t.co` is also used to check malicious domains, so not sure if we should address this issue or not.

What do you guys think? Or maybe we can move it to annoyance list?

**Comment by Yuki2718** [comment]
Yeah, maybe safer to move to annoyance.

**Comment by gwarser** [comment]
Ordinary links cannot be checked for malware?

**Comment by stephenhawk8054** [comment]
> Ordinary links cannot be checked for malware? 

Hmm... By what method?

**Comment by gwarser** [comment]
I think this is just Twitter's excuse to track clicks.

**Comment by Yuki2718** [comment]
Yeah, I haven't ever warned when clicking bad links for badware. But who knows.

**Comment by Yuki2718** [comment]
Found an example Twitter/X warns if via `t.co`: `https://x.com/2cheromatome/status/1508499176874815492` Not actually a dangerous link though.

**Comment by Yuki2718** [comment]
On mobile links in `threads.net` redirects to `l.threads.net` when you loaded any pages, but it self-fixed to actual links if you wait a minute so not sure if this should be addressed.

**Comment by stephenhawk8054** [comment]
I think it's fine to address

**Comment by gorhill** [comment]
> server gets the request it check whether the `redirect_site` is in the list of there malicious(linkshim) list or not

If the URL of a clickable link is malware, why is Facebook allowing these links to be embedded on their site in the first place?

**Comment by dimisa-RUAdList** [comment]
@gorhill 
When using uBlock Origin 1.61.3rc0, some `href-sanitizer` rules stopped working, for example here: `https://vz.ru/news/2024/12/25/1305553.html`
```
vz.ru##+js(href-sanitizer, a[href*="/redir/"], ?exturl)
vz.ru##+js(href-sanitizer, a[href*="/redir/"], ?vzurl)
```
In the stable version of uBlock Origin 1.61.2, everything is fine.

**Comment by gorhill** [comment]
Seems to work from here:

![image](https://github.com/user-attachments/assets/1e1effb1-e251-474e-aa68-afd47f9b3f28)

Which exact link on that page which should be sanitized is not being sanitized?

**Comment by dimisa-RUAdList** [comment]
Yes, these are the links in question. Can this be fixed?

**Comment by gorhill** [comment]
Fixed in next dev build.

**Comment by dimisa-RUAdList** [comment]
Yes, everything is great in uBlock Origin 1.61.3rc1!

**Comment by dimisa-RUAdList** [comment]
Is it possible to add an `+https `option to `href-sanitizer` so that it would be possible to fix links when they point to `http`?

Example: `https://news.rambler.ru/starlife/55642377-glava-fpbk-borodin-vystupil-protiv-chlenstva-dibrova-v-akademii-televideniya/`

<details>
<summary>Screenshot(s)</summary>

![Image](https://github.com/user-attachments/assets/dde1bb09-38e4-4a1e-b1ab-15a72f456353)

</details>

**Comment by gorhill** [comment]
You can already use:

    news.rambler.ru##+js(href-sanitizer, a[href^="http:"], +https)

**Comment by dimisa-RUAdList** [comment]
Great! I didn't know this already worked.
